x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/physical/local/EnableSpatialDistancePushdown.java

```diff
@@ -94,46 +95,82 @@ private FilterExec rewrite(FilterExec filterExec, EsQueryExec esQueryExec) {
         return filterExec;
     }
 
-    private Map<NameId, StDistance> getPushableDistances(List<Alias> aliases) {
-        LinkedHashMap<NameId, StDistance> refs = new LinkedHashMap<>();
+    private void getPushableDistances(List<Alias> aliases, Map<NameId, StDistance> distances, Map<NameId, Alias> others) {
         aliases.forEach(alias -> {
-            if (alias.child() instanceof StDistance distance && canPushSpatialFunctionToSource(distance, Map.of())) {
-                refs.put(alias.id(), distance);
-            } else if (alias.child() instanceof ReferenceAttribute ref && refs.containsKey(ref.id())) {
-                StDistance distance = refs.get(ref.id());
-                refs.put(alias.id(), distance);
+            if (alias.child() instanceof StDistance distance && canPushSpatialFunctionToSource(distance)) {
+                distances.put(alias.id(), distance);
+            } else if (alias.child() instanceof ReferenceAttribute ref && distances.containsKey(ref.id())) {
+                StDistance distance = distances.get(ref.id());
+                distances.put(alias.id(), distance);
+            } else {
+                others.put(alias.id(), alias);
             }
         });
-        return refs;
     }
 
     private PhysicalPlan rewrite(FilterExec filterExec, EvalExec evalExec, EsQueryExec esQueryExec) {
-        Map<NameId, StDistance> pushable = getPushableDistances(evalExec.fields());
-        if (pushable.isEmpty() == false) {
+        Map<NameId, StDistance> distances = new LinkedHashMap<>();
+        Map<NameId, Alias> others = new LinkedHashMap<>();
+        getPushableDistances(evalExec.fields(), distances, others);
+        if (distances.isEmpty() == false) {
             // Find and rewrite any binary comparisons that involve a distance function and a literal
             var rewritten = filterExec.condition().transformDown(EsqlBinaryComparison.class, comparison -> {
                 ComparisonType comparisonType = ComparisonType.from(comparison.getFunctionType());
-                if (comparison.left() instanceof ReferenceAttribute r && pushable.containsKey(r.id()) && comparison.right().foldable()) {
-                    StDistance dist = pushable.get(r.id());
+                if (comparison.left() instanceof ReferenceAttribute r && distances.containsKey(r.id()) && comparison.right().foldable()) {
+                    StDistance dist = distances.get(r.id());
                     return rewriteComparison(comparison, dist, comparison.right(), comparisonType);
                 } else if (comparison.right() instanceof ReferenceAttribute r
```

Comment:
The fact that we rewrite only in case of reference attributes looks incorrect. I think in an execution plan like
```
Filter[st_distance(field{f}, POINT(0 0)) < 10, st_distance(renamed_field{r}, POINT(1 -1)) > 5]
\_Eval[renamed_field = other_field{f}]
  \_Query
```
this will rewrite only the predicate on the `renamed_field` but leave the non-renamed predicate on `field` alone.

I also do not believe that the best approach to deal with aliases that only rename is to keep them before the filter and splitting the eval. I believe we should split the filter and keep the eval alone.

In more detail: let's update the filter by replacing reference attributes by field attributes - then part of the filter can be brought closer to the source. (In fact, that's something that should happen in the logical planning, but let's keep it locally here for this PR and move out in a follow-up.)

The reason why I think that's important is that bringing filters close to the source is something we do in general, and it allows us to focus on combinations `Query -> Filter` for Lucene pushdowns. I'd like to have as little special casing as possible for cases like `Query -> Eval -> Filter`, because looking for patterns consisting of 3 plan nodes is less general and more brittle than looking for patterns consisting of 2 plan nodes.

For example, I think the plan from above should be rewritten to
```
Filter[st_distance(field{f}, POINT(0 0)) < 10, st_distance(other_field{f}, POINT(1 -1)) > 5]
\_Eval[renamed_field = other_field{f}]
  \_Query
```
and then the filter can even be wholly pushed down
```
Eval[renamed_field = other_field{f}]
\_Filter[st_distance(field{f}, POINT(0 0)) < 10, st_distance(other_field{f}, POINT(1 -1)) > 5]
  \_Query
```
So that we only need to replace the `st_distance`s and then `PushFiltersToSource` can just look at the `Query -> Filter` pattern at the bottom and push the filter further down.

In contrast, the rule currently doesn't seem to rewrite this plan at all because the eval is needed for the filter - and thus `PushFiltersToSource` needs to look for the `Query -> Eval -> Filter` pattern.

Reply:
Never mind. Filters that don't have reference attributes that refer to expressions defined in the Eval should have been pushed down past the Eval in the first place, so the situation I described doesn't need to be handled here.